api/src/services/application-csv-export.service.ts

```diff
@@ -815,15 +698,380 @@ export class ApplicationCsvExporterService
+        queryParams.listingId,
+      );
+
+    // get maxHouseholdMembers associated to the selected applications
+    let maxHouseholdMembers = 0;
+    applications.forEach((app) => {
+      if (app.householdMember?.length > maxHouseholdMembers) {
+        maxHouseholdMembers = app.householdMember.length;
+      }
+    });
+
+    const csvHeaders = await this.getCsvHeaders(
+      maxHouseholdMembers,
+      multiSelectQuestions,
+      queryParams.timeZone,
+      queryParams.includeDemographics,
+      true,
+    );
+
+    return this.csvExportHelper(
+      filename,
+      mapTo(Application, applications),
+      csvHeaders,
+      queryParams,
+      true,
+    );
+  }
+
+  /**
+   *
+   * @param queryParams
+   * @param req
+   * @returns generates the lottery results for a listing
+   */
+  async lotteryGenerate<QueryParams extends ApplicationCsvQueryParams>(
+    req: ExpressRequest,
+    res: Response,
+    queryParams: QueryParams,
+  ): Promise<SuccessDTO> {
+    const user = mapTo(User, req['user']);
+    await this.authorizeCSVExport(user, queryParams.listingId);
+
+    const listing = await this.prisma.listings.findUnique({
+      select: {
+        id: true,
+        lotteryLastRunAt: true,
+        lotteryStatus: true,
+      },
+      where: {
+        id: queryParams.listingId,
+      },
+    });
+    if (listing?.lotteryStatus) {
+      // if lottery has been run before
+      throw new BadRequestException(
+        `Listing ${queryParams.listingId}: the lottery was attempted to be generated but it was already run previously`,
+      );
+    }
+
+    const applications = await this.prisma.applications.findMany({
+      select: {
+        id: true,
+        preferences: true,
+        householdMember: {
+          select: {
+            id: true,
+          },
+        },
+        applicationLotteryPositions: {
+          select: {
+            ordinal: true,
+            multiselectQuestionId: true,
+          },
+          where: {
+            multiselectQuestionId: null,
+          },
+          orderBy: {
+            ordinal: OrderByEnum.DESC,
+          },
+        },
+      },
+      where: {
+        listingId: queryParams.listingId,
+        deletedAt: null,
+        markedAsDuplicate: false,
+      },
+    });
+
+    // get all multiselect questions for a listing to build csv headers
+    const multiSelectQuestions =
+      await this.multiselectQuestionService.findByListingId(
+        queryParams.listingId,
+      );
+
+    await this.lotteryRandomizer(
+      queryParams.listingId,
+      mapTo(Application, applications),
+      multiSelectQuestions.filter(
+        (multiselectQuestion) =>
+          multiselectQuestion.applicationSection ===
+          MultiselectQuestionsApplicationSectionEnum.preferences,
+      ),
+    );
+
+    await this.prisma.listings.update({
+      data: {
+        lotteryLastRunAt: new Date(),
+        lotteryStatus: LotteryStatusEnum.ran,
```

Comment:
We also created a LotteryStatusEnum of "errored". Should this function be wrapping everything above this listings update in a try catch function where we can update the lotteryStatus to errored?

Reply:
I would also consider calling the listing.service.lotteryStatus function here so the logic isn't duplicated and we only have to implement the lottery history logic in one location